### PR TITLE
Make crictl tests host local

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
We now use a host local exec instead of SSH commands to simplify the test and make the result more robust.

#### Which issue(s) this PR fixes:
Follow-up of https://github.com/kubernetes/kubernetes/pull/101866

#### Special notes for your reviewer:
None

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
None
```

### CODE REVIEW

**Comment by saschagrunert** [comment]
Test passed:
```
I0712 10:45:36.975] [BeforeEach] [sig-node] crictl
I0712 10:45:36.975]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:185
I0712 10:45:36.975] STEP: Creating a kubernetes client
I0712 10:45:36.976] Jul 12 10:44:47.733: INFO: >>> kubeConfig: /workspace/.kube/config
I0712 10:45:36.976] STEP: Building a namespace api object, basename crictl
I0712 10:45:36.976] STEP: Binding the e2e-test-privileged-psp PodSecurityPolicy to the default service account in crictl-4856
I0712 10:45:36.976] STEP: Waiting for a default service account to be provisioned in namespace
I0712 10:45:36.976] [BeforeEach] [sig-node] crictl
I0712 10:45:36.976]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/node/crictl.go:33
I0712 10:45:36.976] [It] should be able to run crictl on the node
I0712 10:45:36.977]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/node/crictl.go:40
I0712 10:45:36.977] STEP: Testing "sudo crictl version" on node "e2e-61fe79dc5f-a7d53-minion-group-3shq" 
I0712 10:45:36.977] Jul 12 10:45:08.372: INFO: ExecWithOptions {Command:[nsenter --mount=/rootfs/proc/1/ns/mnt -- sh -c sudo crictl version] Namespace:crictl-4856 PodName:hostexec-e2e-61fe79dc5f-a7d53-minion-group-3shq-jf4bz ContainerName:agnhost-container Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
I0712 10:45:36.977] Jul 12 10:45:08.372: INFO: >>> kubeConfig: /workspace/.kube/config
I0712 10:45:36.977] STEP: Testing "sudo crictl version" on node "e2e-61fe79dc5f-a7d53-minion-group-czd6" 
I0712 10:45:36.978] Jul 12 10:45:21.987: INFO: ExecWithOptions {Command:[nsenter --mount=/rootfs/proc/1/ns/mnt -- sh -c sudo crictl version] Namespace:crictl-4856 PodName:hostexec-e2e-61fe79dc5f-a7d53-minion-group-czd6-d52jg ContainerName:agnhost-container Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
I0712 10:45:36.978] Jul 12 10:45:21.987: INFO: >>> kubeConfig: /workspace/.kube/config
I0712 10:45:36.978] STEP: Testing "sudo crictl version" on node "e2e-61fe79dc5f-a7d53-minion-group-wdqg" 
I0712 10:45:36.979] Jul 12 10:45:34.717: INFO: ExecWithOptions {Command:[nsenter --mount=/rootfs/proc/1/ns/mnt -- sh -c sudo crictl version] Namespace:crictl-4856 PodName:hostexec-e2e-61fe79dc5f-a7d53-minion-group-wdqg-ds45h ContainerName:agnhost-container Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
I0712 10:45:36.979] Jul 12 10:45:34.717: INFO: >>> kubeConfig: /workspace/.kube/config
I0712 10:45:36.979] STEP: Testing "sudo crictl info" on node "e2e-61fe79dc5f-a7d53-minion-group-3shq" 
I0712 10:45:36.979] Jul 12 10:45:35.399: INFO: ExecWithOptions {Command:[nsenter --mount=/rootfs/proc/1/ns/mnt -- sh -c sudo crictl info] Namespace:crictl-4856 PodName:hostexec-e2e-61fe79dc5f-a7d53-minion-group-3shq-jf4bz ContainerName:agnhost-container Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
I0712 10:45:36.979] Jul 12 10:45:35.399: INFO: >>> kubeConfig: /workspace/.kube/config
I0712 10:45:36.980] STEP: Testing "sudo crictl info" on node "e2e-61fe79dc5f-a7d53-minion-group-czd6" 
I0712 10:45:36.980] Jul 12 10:45:35.829: INFO: ExecWithOptions {Command:[nsenter --mount=/rootfs/proc/1/ns/mnt -- sh -c sudo crictl info] Namespace:crictl-4856 PodName:hostexec-e2e-61fe79dc5f-a7d53-minion-group-czd6-d52jg ContainerName:agnhost-container Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
I0712 10:45:36.980] Jul 12 10:45:35.829: INFO: >>> kubeConfig: /workspace/.kube/config
I0712 10:45:36.980] STEP: Testing "sudo crictl info" on node "e2e-61fe79dc5f-a7d53-minion-group-wdqg" 
I0712 10:45:36.981] Jul 12 10:45:36.346: INFO: ExecWithOptions {Command:[nsenter --mount=/rootfs/proc/1/ns/mnt -- sh -c sudo crictl info] Namespace:crictl-4856 PodName:hostexec-e2e-61fe79dc5f-a7d53-minion-group-wdqg-ds45h ContainerName:agnhost-container Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
I0712 10:45:36.981] Jul 12 10:45:36.346: INFO: >>> kubeConfig: /workspace/.kube/config
I0712 10:45:36.981] [AfterEach] [sig-node] crictl
I0712 10:45:36.981]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:186
I0712 10:45:36.981] Jul 12 10:45:36.843: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
I0712 10:45:36.982] STEP: Destroying namespace "crictl-4856" for this suite.
I0712 10:45:36.982]
I0712 10:45:36.982]
I0712 10:45:36.982] [SLOW TEST:49.241 seconds]
I0712 10:45:36.982] [sig-node] crictl
I0712 10:45:36.982] /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/node/framework.go:23
I0712 10:45:36.982]   should be able to run crictl on the node
I0712 10:45:36.982]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/node/crictl.go:40
I0712 10:45:36.982] ------------------------------
I0712 10:45:36.983] {"msg":"PASSED [sig-node] crictl should be able to run crictl on the node","total":-1,"completed":8,"skipped":24,"failed":0}
```

**Comment by saschagrunert** [comment]
/override pull-kubernetes-kubemark-e2e-gce-big

I don't think we have to rely on those because `pull-kubernetes-e2e-gce-ubuntu-containerd` already runs the modified test.

**Comment by endocrimes** [comment]
/lgtm for the change itself

I think that should be true for the scalability tests (with the caveat that I am fairly new to them)

**Review by feiskyer — APPROVED** [review]
/lgtm

File: test/e2e/node/crictl.go
Comment (SergeyKanzhelev) [inline test/e2e/node/crictl.go:41]: it used to skip the test if there are no ready ssh-able nodes, with this change it will fail. Is it intentional? (PR introduced test skipping: https://github.com/kubernetes/kubernetes/pull/101866)

Comment (oomichi) [reply test/e2e/node/crictl.go:41]: This new combination of `GetBoundedReadySchedulableNodes()` and `hostExec.Execute()` creates a pod on the specific node and run the specified command in the pod without using ssh.
This pr requires just schedulable nodes, and I think it is good that the test fails if there is not such nodes on the environment.

File: test/e2e/node/crictl.go
Comment (oomichi) [inline test/e2e/node/crictl.go:59]: just a question: Does this new string length check solve the original TODO of the original line 66?
I am not sure what was the intention of the original TODO.

File: test/e2e/node/crictl.go
Comment (saschagrunert) [reply test/e2e/node/crictl.go:41]: Yeah the test is also still covered by the e2eskipper constraint:

```
ginkgo.BeforeEach(func() {
		// `crictl` is not available on all cloud providers.
		e2eskipper.SkipUnlessProviderIs("gce", "gke")
	})
```

File: test/e2e/node/crictl.go
Comment (saschagrunert) [reply test/e2e/node/crictl.go:57]: I think so. I was considering to test the output directly but depending on the `crictl` version it may differ. Getting anything on stdout seems fine from my point of view for now.

**Comment by saschagrunert** [comment]
I'm not sure if the suites are the same but since `ci-kubernetes-e2e-gce-scale-correctness` seems to fail after my previous merge:
/test pull-kubernetes-kubemark-e2e-gce-scale

**Comment by saschagrunert** [comment]
Does not look like that it triggers the test
/override pull-kubernetes-kubemark-e2e-gce-scale

**Comment by ehashman** [comment]
Looks like the test is still failing against this PR: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103640/pull-kubernetes-e2e-gce-correctness/1416140241235873792 :confused: 

/sig scalability
FYI

File: test/e2e/node/ssh.go
Comment (saschagrunert) [inline test/e2e/node/ssh.go:52]: This fixes the SSH node test and skips the test if no SSH'able nodes are available.

**Comment by saschagrunert** [comment]
Closes https://github.com/kubernetes/kubernetes/issues/103688

File: test/e2e/node/crictl.go
Comment (aojea) [reply test/e2e/node/crictl.go:39]: where is `maxNodes`coming from?

File: test/e2e/node/crictl.go
Comment (aojea) [inline test/e2e/node/crictl.go:25]: side note, @oomichi we should start thinking in moving the `NewHostExec` to a more general location, right?
I start to see it in more places than in storage:
```
test/e2e/node/mount_propagation.go:             hostExec := utils.NewHostExec(f)
test/e2e/storage/utils/host_exec.go:// NewHostExec returns a HostExec
test/e2e/storage/utils/host_exec.go:func NewHostExec(framework *framework.Framework) HostExec {
test/e2e/storage/persistent_volumes-local.go:           hostExec := utils.NewHostExec(f)
test/e2e/storage/testsuites/volumemode.go:              hostExec := storageutils.NewHostExec(f)
test/e2e/storage/testsuites/subpath.go:         l.hostExec = utils.NewHostExec(f)
test/e2e/storage/drivers/in_tree.go:    l.hostExec = utils.NewHostExec(f)
test/e2e/network/netpol/network_legacy.go:      hostExec := utils.NewHostExec(f)
test/e2e/network/util.go:       hostExec := utils.NewHostExec(f)
test/e2e/network/service.go:            hostExec := utils.NewHostExec(f)
test/e2e/network/service.go:            hostExec := utils.NewHostExec(f)
```
cc: @BenTheElder

File: test/e2e/node/crictl.go
Comment (aojea) [reply test/e2e/node/crictl.go:39]: Is there a relation of `gce,gke`and `crictl` or just the CI combination?
I think that this will run now everywhere where the binary is available, i.e., it will run on kind jobs too, 

TODO to skip if binary doesn't exist and not skip by cloud provider

File: test/e2e/node/crictl.go
Comment (aojea) [inline test/e2e/node/crictl.go:44]: no need to sudo since the hostExec pod runs privileged and some OS doesn't have sudo (i.e. kind jobs)

https://github.com/kubernetes/kubernetes/blob/c82626f96f754e0eb87eaa1380bb672be42e6f0d/test/e2e/storage/utils/host_exec.go#L102-L106

File: test/e2e/node/ssh.go
Comment (spiffxp) [reply test/e2e/node/ssh.go:52]: But by this point shouldn't we expect that there are SSH'able nodes and consider it an error if there are none?

/cc @wojtek-t @mm4tt
Would like input from sig-scalability as this is being point to as a fix for https://github.com/kubernetes/kubernetes/issues/103688

File: test/e2e/node/crictl.go
Comment (spiffxp) [inline test/e2e/node/crictl.go:36]: Not commenting on the removals but the line above.  Is there some less-tied-to-cluster/-way of knowing whether `crictl` is available?

**Review by spiffxp — CHANGES_REQUESTED** [review]
Use hostexec for the `crictl` test looks valid to me, though I wish there was a better way than 'provider' of determining whether `crictl` is present (how do you run these tests on something stood up by not-kube-up.sh?)

Skipping instead of failing tests that literally have the word "SSH" in their description seems incorrect though

File: test/e2e/node/crictl.go
Comment (spiffxp) [reply test/e2e/node/crictl.go:36]: (crictl doesn't seem like a cloud-provider implementation detail)

File: test/e2e/node/ssh.go
Comment (spiffxp) [reply test/e2e/node/ssh.go:52]: Remove this line and I'm happy to approve if SIG Node is happy with it

Comment (marseel) [reply test/e2e/node/ssh.go:52]: I would prefer to remove this line as well. I tried to understand reason behind failing this test in scalability job and not in regular tests. The difference is that in scalability jobs we don't have public IPs on our nodes and we use "gcp-ssh-proxy-instance-name" kubetes flag. This flag sets "KUBE_SSH_BASTION" env.  You can see here that there is special case for ssh when this env is set: https://github.com/kubernetes/kubernetes/blob/e3363152383ce8fc210290929e366ffdcbf71af9/test/e2e/framework/ssh/ssh.go#L210 
So I believe that we should take this into account when checking for list of available hosts here: https://github.com/kubernetes/kubernetes/blob/fac3dd6914f95e57975e7de7ccb3de2777f9edd6/test/e2e/framework/ssh/ssh.go#L143

File: test/e2e/framework/ssh/ssh.go
Comment (wojtek-t) [inline test/e2e/framework/ssh/ssh.go:159]: Looking into: https://github.com/kubernetes/kubernetes/pull/101866 - this clearly broke scalability correctness tests.

From this line in the log file:
```
ï¿½[1mSTEPï¿½[0m: SSH'ing to 0 nodes and running echo "Hello from $(whoami)@$(hostname)"
```

We know that after that PR, the function returns 0 nodes (where it used to return at least 1).

We should fix the `canConnect` function.

File: test/e2e/node/ssh.go
Comment (wojtek-t) [reply test/e2e/node/ssh.go:52]: Agree - that's an incorrect change, we don't want this.
See this comment: https://github.com/kubernetes/kubernetes/pull/103640/files#r673923883

**Comment by saschagrunert** [comment]
Decided to move the SSH host fix out of scope of this PR and debugging the issue in https://github.com/kubernetes/kubernetes/pull/103815

File: test/e2e/node/crictl.go
Comment (saschagrunert) [reply test/e2e/node/crictl.go:36]: Hm, how would we do that? Could we say every node should have `crictl` configured? (tending to no)

File: test/e2e/node/crictl.go
Comment (saschagrunert) [reply test/e2e/node/crictl.go:25]: Yeah I thought the same. I think I could follow-up on that and move it into a more dedicated location.

File: test/e2e/node/crictl.go
Comment (spiffxp) [reply test/e2e/node/crictl.go:36]: Honestly if crictl can't reasonably be expected to exist by default on all clusters, this should be tagged with [Feature:CrictlInstalled] or something. I'll leave it be since this is how this has always been, but we should reconsider this for v1.23 or if we ever want to untangle from kube-up.sh

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

File: test/e2e/node/crictl.go
Comment (aojea) [reply test/e2e/node/crictl.go:25]: yeah, follow up, not drive by, just saw it and added a comment so I don't forget

**Comment by ehashman** [comment]
/milestone clear

from SIG Node CI subgroup today, this isn't release-critical

**Comment by dims** [comment]
/approve
/lgtm

(re-applying lgtm lost in rebase from multiple folks as well)

**Pull request merged**
